pkg/controller/upgradeconfig/upgradeconfig_controller_test.go

```diff
@@ -1,74 +1,417 @@
+					It("Invokes the upgrader", func() {
+						mockClusterUpgrader.EXPECT().UpgradeCluster(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Times(1)
+						result, err := reconciler.Reconcile(reconcile.Request{NamespacedName: upgradeConfigName})
+						Expect(err).NotTo(HaveOccurred())
+						Expect(result.Requeue).To(BeFalse())
+						Expect(result.RequeueAfter).To(BeZero())
+					})
+				})
+			})
+
+			Context("When the upgrade phase is Pending", func() {
+				JustBeforeEach(func() {
+					upgradeConfig.Status.History[0].Phase = upgradev1alpha1.UpgradePhasePending
+				})
+
+				Context("When the cluster is ready to upgrade", func() {
+					Context("When a maintenance client can't be built", func() {
+						var fakeError = fmt.Errorf("a maintenance builder error")
+						JustBeforeEach(func() {
+							maintenanceBuilder := func(client client.Client) (maintenance.Maintenance, error) {
+								return mockMaintenanceClient, fakeError
+							}
+							reconciler.maintenanceClientBuilder = maintenanceBuilder
+						})
+						It("does not proceed with upgrading the cluster", func() {
+							mockClusterUpgrader.EXPECT().UpgradeCluster(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Times(0)
+							result, err := reconciler.Reconcile(reconcile.Request{NamespacedName: upgradeConfigName})
+							Expect(err).To(Equal(fakeError))
+							Expect(result.Requeue).To(BeFalse())
+							Expect(result.RequeueAfter).To(BeZero())
+						})
+					})
+
+					Context("When a maintenance client can be built", func() {
+						It("Invokes the upgrader", func() {
+							mockClusterUpgrader.EXPECT().UpgradeCluster(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Times(1)
+							result, err := reconciler.Reconcile(reconcile.Request{NamespacedName: upgradeConfigName})
+							Expect(err).NotTo(HaveOccurred())
+							Expect(result.Requeue).To(BeFalse())
+							Expect(result.RequeueAfter).To(BeZero())
+						})
+					})
+
+					Context("When invoking the upgrader fails", func() {
+						var fakeError = fmt.Errorf("the upgrader failed")
+						It("reacts accordingly", func() {
+							// All it does here is log at the moment
+							mockClusterUpgrader.EXPECT().UpgradeCluster(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Times(1).Return(fakeError)
+							result, err := reconciler.Reconcile(reconcile.Request{NamespacedName: upgradeConfigName})
+							Expect(err).NotTo(HaveOccurred())
+							Expect(result.Requeue).To(BeFalse())
+							Expect(result.RequeueAfter).To(BeZero())
+						})
+					})
+				})
+
+				Context("When the cluster is not ready to upgrade", func() {
+					// TODO: This is never true at the moment - no readiness check implemented
+				})
+			})
+
+			Context("When the upgrade phase is Upgrading", func() {
+				JustBeforeEach(func() {
+					upgradeConfig.Status.History[0].Phase = upgradev1alpha1.UpgradePhaseUpgrading
+				})
+
+				Context("When a maintenance client can't be built", func() {
```

Comment:
Might review is some of these make more sense as cluster_upgrader tests rather than controller tests if we create a new pkg.

Reply:
Can agree it might feel a bit fuzzy, but every test in this file was needed to ensure that every bit of code in the reconcile function gets coverage. I think there may be cases where some test is valid for both pkgs.